utility/events.lua

```diff
@@ -0,0 +1,571 @@
+        -- ... register its dispatch logic with Factorio
+        script.on_configuration_changed(function(event_data)
+            for _, on_configuration_changed_handler in ipairs(on_configuration_changed_handlers) do
+                on_configuration_changed_handler(event_data)
+            end
+        end)
+    end
+
+    -- Register the handler
+    table.insert(on_configuration_changed_handlers, handler)
+end
+
+--#endregion
+
+--#region on_event
+
+---Subscribers' handlers for on_event
+---@type Dictionary<LuaEventType, Array<fun(event_data: EventData)>>
+local on_event_handlers = {}
+
+-- TODO: Open issue for LuaCATS to support callback parameter annotations.
+--       E.g. for `event_data` below, related to the FIXME above.
+
+-- TODO: To support filters in the future.
+--       The filter passed with each handler must be stored separately and
+--       the dispatcher must be reregistered with the superset of all filters.
+
+---Register a handler to run on the specified event(s).
+---@param event LuaEventType|Array<LuaEventType> The event(s) or custom-input to invoke the handler on.
+---@param handler fun(event_data: EventData) The handler for this event.
+---
+---Wraps `script.on_event`, allowing multiple handlers to be registered for the same event types without overwriting each other.
+---Passing `nil` to unregister is not supported. (A new concept for unregistering handlers may be added in a future version.)
+---Filters are not supported yet, but are planned for a future version.
+---
+---**Important:** Due to how Factorio events work, calling `script.on_event` directly for an event handled by `events.on_event` will interfere with `events.on_event`, disabling it for that event.
+---
+---### Example
+---
+---```
+----- Register for the on_tick event to print the current tick to console each tick
+---events.on_event(defines.events.on_tick, function(event_data)
+---  game.print("Current tick: " .. event_data.tick)
+---end)
+---```
+---
+---### Original Description
+---
+---Register a handler to run on the specified event(s). Each mod can only register once for every event, as any additional registration will overwrite the previous one. This holds true even if different filters are used for subsequent registrations.
+---
+---### Example
+---
+---```
+----- Register for the on_tick event to print the current tick to console each tick
+---script.on_event(defines.events.on_tick,
+---function(event) game.print(event.tick) end)
+---```
+---
+---### Example
+---
+---```
+----- Register for the on_built_entity event, limiting it to only be received when a `"fast-inserter"` is built
+---script.on_event(defines.events.on_built_entity,
+---function(event) game.print("Gotta go fast!") end,
+---{{filter = "name", name = "fast-inserter"}})
+---```
+---
+---[View Documentation](https://lua-api.factorio.com/latest/classes/LuaBootstrap.html#on_event)
+---
+---@overload fun(event: string, handler: fun(event_data: EventData))
```

Comment:
This has some code smell. While I see the appeal of using this to give type hinting, this means that we will need to make a new commit every time the Factorio devs release a change that affects any of these events.

While I'm generally not against something like this, this feels like something that is better lived in its own library mod, similar to https://mods.factorio.com/mod/flib, which can then be used as a dependency for mods. That said, this would potentially cause issues as registering functions into a different mod makes it more complicated to unregister events in the future, so I'm not sure if it's something that's desirable.

Reply:
Each mod gets its own sandbox, if the same file from the same mod is required by two different mods, it would result in two separate independent tables, so unregistering events, or consuming the module at all would not change, independently of the mod it resides in.

Regarding the concern about updating this repository each time Factorio developers introduce or modify events, that would only be necessary if we intend to keep all the latest APIs available indefinitely, even when we do not actively use them. If we do use new features, we would need to update and commit anyway.

Additional thoughts/opinions on this:
- Changes that require updates here would likely be breaking changes. Factorio’s development is essentially winding down, so I do not anticipate many of these. If they occur, they will probably happen by version 2.1., which is supposed to be the last major update.
- New events do not need to be added unless they are directly relevant to this mod.
- Not adding the necessary overloads will cause type mismatch warnings from the language server.
- I plan to introduce events as separate functions over time. Whenever a new function is added, I will remove the corresponding overload from on_event, thereby encouraging users to adopt the dedicated event functions. Ultimately, I aim to avoid using on_event entirely.
- This should serve as a library or module, and I am considering distributing it via LuaRocks or as a separate mod some time later.
  - I prefer not to rely on large Factorio libraries containing many unused features that require constant updates. Instead, I favor a modular approach, similar to .NET Core, where you select only the modules needed.
  - Embedding a small, easily updatable module within a mod is practical. Should maintenance cease, users can simply update the parts they rely on, preserving functionality.
  - Once events.lua is reused in multiple mods, I will likely release it as a LuaRocks package.
  - I prefer the LuaRocks approach rather than creating a separate mod, since it simplifies integration with language servers and avoids forcing all consumers to adopt the same version.

Do you see any way to avoid including these overloads without triggering language server warnings, while still ensuring that incompatible API usage produces the appropriate warnings?